src/bench/connectblock.cpp

```diff
@@ -0,0 +1,145 @@
+
+    std::vector<CMutableTransaction> txs;
+    txs.reserve(num_txs);
+    CTransactionRef input_tx{MakeTransactionRef(first_tx)};
+    for (int i{0}; i < num_txs; i++) {
+        std::vector<COutPoint> inputs;
+        inputs.reserve(outputs.size());
+
+        for (size_t j{0}; j < outputs.size(); j++) {
+            inputs.emplace_back(input_tx->GetHash(), j);
+        }
+
+        const auto [taproot_tx, _]{test_setup.CreateValidTransaction(
+            std::vector{input_tx}, inputs, chainstate.m_chain.Height() + 1, keys, outputs, std::nullopt, std::nullopt)};
+        txs.emplace_back(taproot_tx);
+        input_tx = MakeTransactionRef(taproot_tx);
+    }
+
+    // Coinbase output can use any output type as it is not spent and will not change the benchmark
+    const CScript coinbase_spk{GetScriptForDestination(coinbase_taproot)};
+    return test_setup.CreateBlock(txs, coinbase_spk, chainstate);
+}
+
+/*
+ * Creates key pairs and corresponding outputs for the benchmark transactions.
+ * - For Taproot outputs: Creates simple key path spendable outputs
+ * - For non-Taproot outputs: Creates P2WPKH (native SegWit v0) outputs
+ * - All outputs have value of 1 BTC
+ */
+std::pair<std::vector<CKey>, std::vector<CTxOut>> CreateKeysAndOutputs(const CKey& coinbaseKey, size_t num_taproot, size_t num_nontaproot)
+{
+    std::vector<CKey> keys{coinbaseKey};
+    keys.reserve(num_taproot + num_nontaproot + 1);
+
+    std::vector<CTxOut> outputs;
+    outputs.reserve(num_taproot + num_nontaproot);
+
+    for (size_t i{0}; i < num_nontaproot; i++) {
+        const CKey key{GenerateRandomKey()};
+        keys.emplace_back(key);
+        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV0KeyHash{key.GetPubKey()}));
+    }
+
+    for (size_t i{0}; i < num_taproot; i++) {
+        CKey key{GenerateRandomKey()};
+        keys.emplace_back(key);
+        outputs.emplace_back(COIN, GetScriptForDestination(WitnessV1Taproot{XOnlyPubKey(key.GetPubKey())}));
+    }
+
+    return {keys, outputs};
+}
+
+void BenchmarkConnectBlock(benchmark::Bench& bench, std::vector<CKey>& keys, std::vector<CTxOut>& outputs, TestChain100Setup& test_setup)
+{
+    const auto test_block{CreateTestBlock(test_setup, keys, outputs)};
+    auto pindex{std::make_unique<CBlockIndex>(test_block)};
+    auto test_blockhash{std::make_unique<uint256>(test_block.GetHash())};
+
+    Chainstate& chainstate{test_setup.m_node.chainman->ActiveChainstate()};
+
+    pindex->nHeight = chainstate.m_chain.Height() + 1;
+    pindex->phashBlock = test_blockhash.get();
+    pindex->pprev = chainstate.m_chain.Tip();
+
+    BlockValidationState test_block_state;
+    bench.unit("block").run([&] {
+        LOCK(cs_main);
+        CCoinsViewCache viewNew{&chainstate.CoinsTip()};
+        assert(chainstate.ConnectBlock(test_block, test_block_state, pindex.get(), viewNew));
+    });
+}
+
+static void ConnectBlockAllSchnorr(benchmark::Bench& bench)
```

Comment:
This is what the flames look like - I haven't worked in this area yet, do you think it's representative?
<img width="1000" alt="image" src="https://github.com/user-attachments/assets/d55f64e2-749f-4b7c-b0fb-4e876f2e35af" />

and

<img width="1000" alt="image" src="https://github.com/user-attachments/assets/d0770f14-d29b-480e-ab78-8f37ab223ff1" />